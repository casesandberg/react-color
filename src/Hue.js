export default from './components/hue/Hue'
