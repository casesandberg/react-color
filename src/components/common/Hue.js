'use strict' /* @flow */

import React from 'react'
import ReactCSS from 'reactcss'
import shallowCompare from 'react-addons-shallow-compare'

export class Hue extends ReactCSS.Component {
  shouldComponentUpdate = shallowCompare.bind(this, this, arguments[0], arguments[1])

  classes(): any {
    return {
      'default': {
        hue: {
          Absolute: '0px 0px 0px 0px',
          background: 'linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)',
          borderRadius: this.props.radius,
          boxShadow: this.props.shadow,
        },
        container: {
          margin: '0px 2px',
          position: 'relative',
          height: '100%',
        },
        pointer: {
          zIndex: '2',
          position: 'absolute',
          left: (this.props.hsl.h * 100) / 360 + '%',
        },
        slider: {
          marginTop: '1px',
          width: '4px',
          borderRadius: '1px',
          height: '8px',
          boxShadow: '0px 0px 2px rgba(0, 0, 0, .6)',
          background: '#fff',
          transform: 'translateX(-2px)',
        },
      },
      'direction-vertical': {
        hue: {
          background: 'linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)',
        },
        pointer: {
          left: '0px',
          top: -((this.props.hsl.h * 100) / 360) + 100 + '%',
        },
      },
    }
  }

  componentWillUnmount() {
    this.unbindEventListeners()
  }

  handleChange = (e: any, skip: boolean) => {
    !skip && e.preventDefault()
    var container = this.refs.container
    var containerWidth = container.clientWidth
    var containerHeight = container.clientHeight
    var left = (e.pageX || e.touches[0].pageX) - (container.getBoundingClientRect().left + window.pageXOffset)
    var top = (e.pageY || e.touches[0].pageY) - (container.getBoundingClientRect().top + window.pageYOffset)

    if (this.props.direction === 'vertical') {
      var h
      if (top < 0) {
        h = 359
      } else if (top > containerHeight) {
        h = 0
      } else {
        var percent = -(top * 100 / containerHeight) + 100
        h = (360 * percent / 100)
      }

      if (this.props.hsl.h !== h) {
        this.props.onChange({
          h: h,
          s: this.props.hsl.s,
          l: this.props.hsl.l,
          a: this.props.hsl.a,
          source: 'rgb',
        })
      }
    } else {
      var h
      if (left < 0) {
        h = 0
      } else if (left > containerWidth) {
        h = 359
      } else {
        var percent = left * 100 / containerWidth
        h = (360 * percent / 100)
      }

      if (this.props.hsl.h !== h) {
        this.props.onChange({
          h: h,
          s: this.props.hsl.s,
          l: this.props.hsl.l,
          a: this.props.hsl.a,
          source: 'rgb',
        })
      }
    }
  }

  handleMouseDown = (e: any) => {
    this.handleChange(e, true)
    window.addEventListener('mousemove', this.handleChange)
    window.addEventListener('mouseup', this.handleMouseUp)
  }

  handleMouseUp = () => {
    this.unbindEventListeners()
  }

  unbindEventListeners() {
    window.removeEventListener('mousemove', this.handleChange)
    window.removeEventListener('mouseup', this.handleMouseUp)
  }

  render(): any {
    var pointer = <div is="slider" />

    if (this.props.pointer) {
      pointer = <this.props.pointer {...this.props} />
    }

    return (
      <div is="hue">
        <div is="container" ref="container" onMouseDown={ this.handleMouseDown }
            onTouchMove={ this.handleChange }
            onTouchStart={ this.handleChange }>
          <div is="pointer" ref="pointer">
            { pointer }
          </div>
        </div>
      </div>
    )
  }
}

export default Hue
