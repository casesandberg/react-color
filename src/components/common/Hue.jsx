'use strict';

var React = require('react');
var ReactCSS = require('reactcss');

class Hue extends ReactCSS.Component {

  constructor() {
    super();

    this.handleChange = this.handleChange.bind(this);
    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);
  }

  classes() {
    return {
      'default': {
        hue: {
          Absolute: '0 0 0 0',
          background: 'linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)',
          borderRadius: this.props.radius,
          boxShadow: this.props.shadow,
        },
        container: {
          margin: '0 2px',
          position: 'relative',
          height: '100%',
        },
        pointer: {
          zIndex: '2',
          position: 'absolute',
          left: (this.props.hsl.h * 100) / 360 + '%',
        },
        slider: {
          marginTop: '1px',
          width: '4px',
          borderRadius: '1px',
          height: '8px',
          boxShadow: '0 0 2px rgba(0, 0, 0, .6)',
          background: '#fff',
          transform: 'translateX(-2px)',
        },
      },
      'direction-vertical': {
        hue: {
          background: 'linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)',
        },
        pointer: {
          left: '0',
          top: -((this.props.hsl.h * 100) / 360) + 100 + '%',
        },
      },
    };
  }

  handleChange(e, skip) {
    !skip && e.preventDefault();
    var container = this.refs.container;
    var containerWidth = container.clientWidth;
    var containerHeight = container.clientHeight;
    var left = (e.pageX || e.touches[0].pageX) - (container.getBoundingClientRect().left + window.pageXOffset);
    var top = (e.pageY || e.touches[0].pageY) - (container.getBoundingClientRect().top + window.pageYOffset);

    if (this.props.direction === 'vertical') {
      var h;
      if (top < 0) {
        h = 359;
      } else if (top > containerHeight) {
        h = 0;
      } else {
        var percent = -(top * 100 / containerHeight) + 100;
        h = (360 * percent / 100);
      }

      if (this.props.hsl.h !== h) {
        this.props.onChange({ h: h, s: this.props.hsl.s, l: this.props.hsl.l, a: this.props.hsl.a });
      }
    } else {
      var h;
      if (left < 0) {
        h = 0;
      } else if (left > containerWidth) {
        h = 359;
      } else {
        var percent = left * 100 / containerWidth;
        h = (360 * percent / 100);
      }

      if (this.props.hsl.h !== h) {
        this.props.onChange({ h: h, s: this.props.hsl.s, l: this.props.hsl.l, a: this.props.hsl.a });
      }
    }
  }

  handleMouseDown(e) {
    this.handleChange(e, true);
    window.addEventListener('mousemove', this.handleChange);
    window.addEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseUp() {
    window.removeEventListener('mousemove', this.handleChange);
    window.removeEventListener('mouseup', this.handleMouseUp);
  }

  render() {
    var pointer = <div is="slider" />;

    if (this.props.pointer) {
      pointer = <this.props.pointer {...this.props} />;
    }

    return (
      <div is="hue">
        <div is="container" ref="container" onMouseDown={ this.handleMouseDown } onTouchMove={ this.handleChange }>
          <div is="pointer" ref="pointer">
            { pointer }
          </div>
        </div>
      </div>
    );
  }
}

module.exports = Hue;
