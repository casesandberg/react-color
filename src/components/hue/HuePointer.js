import React from 'react'
import reactCSS from 'reactcss'

export const SliderPointer = ({ direction }) => {
  const styles = reactCSS({
    'default': {
      picker: {
        width: '18px',
        height: '18px',
        borderRadius: '50%',
        transform: 'translate(-9px, -1px)',
        backgroundColor: 'rgb(248, 248, 248)',
        boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
      },
    },
    'vertical': {
      picker: {
        transform: 'translate(-3px, -9px)',
      },
    },
  }, { vertical: direction === 'vertical' })

  return (
    <div style={ styles.picker } />
  )
}

export default SliderPointer
