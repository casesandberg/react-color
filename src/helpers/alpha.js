export const calculateChange = (e, skip, props, container) => {
  e.preventDefault()
  const containerWidth = container.clientWidth
  const containerHeight = container.clientHeight
  const x = typeof e.pageX === 'number' ? e.pageX : e.touches[0].pageX
  const y = typeof e.pageY === 'number' ? e.pageY : e.touches[0].pageY
  const left = x - (container.getBoundingClientRect().left + window.pageXOffset)
  const top = y - (container.getBoundingClientRect().top + window.pageYOffset)

  if (props.direction === 'vertical') {
    let a
    if (top < 0) {
      a = 0
    } else if (top > containerHeight) {
      a = 1
    } else {
      a = Math.round((top * 100) / containerHeight) / 100
    }

    if (props.hsl.a !== a) {
      return {
        h: props.hsl.h,
        s: props.hsl.s,
        l: props.hsl.l,
        a,
        source: 'rgb',
      }
    }
  } else {
    let a
    if (left < 0) {
      a = 0
    } else if (left > containerWidth) {
      a = 1
    } else {
      a = Math.round((left * 100) / containerWidth) / 100
    }

    if (props.a !== a) {
      return {
        h: props.hsl.h,
        s: props.hsl.s,
        l: props.hsl.l,
        a,
        source: 'rgb',
      }
    }
  }
  return null
}
