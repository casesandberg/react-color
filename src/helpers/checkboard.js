const checkboardCache = {}

export const render = (c1, c2, size, serverCanvas) => {
  if (typeof document === 'undefined' && !serverCanvas) {
    return null
  }
  const canvas = serverCanvas ? new serverCanvas() : document.createElement('canvas')
  canvas.width = size * 2
  canvas.height = size * 2
  const ctx = canvas.getContext('2d')
  if (!ctx) {
    return null
  } // If no context can be found, return early.
  ctx.fillStyle = c1
  ctx.fillRect(0, 0, canvas.width, canvas.height)
  ctx.fillStyle = c2
  ctx.fillRect(0, 0, size, size)
  ctx.translate(size, size)
  ctx.fillRect(0, 0, size, size)
  return canvas.toDataURL()
}

export const get = (c1, c2, size, serverCanvas) => {
  const key = `${ c1 }-${ c2 }-${ size }${ serverCanvas ? '-server' : '' }`

  if (checkboardCache[key]) {
    return checkboardCache[key]
  }

  const checkboard = render(c1, c2, size, serverCanvas)
  checkboardCache[key] = checkboard
  return checkboard
}
