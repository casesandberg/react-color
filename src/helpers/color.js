import each from 'lodash/each'
import tinycolor from 'tinycolor2'

export default {
  simpleCheckForValidColor(data) {
    const keysToCheck = ['r', 'g', 'b', 'a', 'h', 's', 'l', 'v']
    let checked = 0
    let passed = 0
    each(keysToCheck, (letter) => {
      if (data[letter]) {
        checked += 1
        if (!isNaN(data[letter])) {
          passed += 1
        }
        if (letter === 's' || letter === 'l') {
          const percentPatt = /^\d+%$/
          if (percentPatt.test(data[letter])) {
            passed += 1
          }
        }
      }
    })
    return (checked === passed) ? data : false
  },

  toState(data, oldHue) {
    const color = data.hex ? tinycolor(data.hex) : tinycolor(data)
    const hsl = color.toHsl()
    const hsv = color.toHsv()
    const rgb = color.toRgb()
    const hex = color.toHex()
    if (hsl.s === 0) {
      hsl.h = oldHue || 0
      hsv.h = oldHue || 0
    }
    const transparent = hex === '000000' && rgb.a === 0

    return {
      hsl,
      hex: transparent ? 'transparent' : `#${ hex }`,
      rgb,
      hsv,
      oldHue: data.h || oldHue || hsl.h,
      source: data.source,
    }
  },

  isValidHex(hex) {
    return tinycolor(hex).isValid()
  },

  getContrastingColor(data) {
    if (!data) {
      return '#fff'
    }
    const col = this.toState(data)
    if (col.hex === 'transparent') {
      return 'rgba(0,0,0,0.4)'
    }
    const yiq = ((col.rgb.r * 299) + (col.rgb.g * 587) + (col.rgb.b * 114)) / 1000
    return (yiq >= 128) ? '#000' : '#fff'
  },
}

export const red = {
  hsl: { a: 1, h: 0, l: 0.5, s: 1 },
  hex: '#ff0000',
  rgb: { r: 255, g: 0, b: 0, a: 1 },
  hsv: { h: 0, s: 1, v: 1, a: 1 },
}
