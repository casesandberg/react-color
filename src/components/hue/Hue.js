import React from 'react'
import reactCSS from 'reactcss'

import { ColorWrap, Hue } from '../common'
import HuePointer from './HuePointer'

export const HuePicker = ({ width, height, onChange, hsl, direction, pointer,
  className = '' }) => {
  const styles = reactCSS({
    'default': {
      picker: {
        position: 'relative',
        width,
        height,
      },
      hue: {
        radius: '2px',
      },
    },
  })

  // Overwrite to provide pure hue color
  const handleChange = data => onChange({ a: 1, h: data.h, l: 0.5, s: 1 })

  return (
    <div style={ styles.picker } className={ `hue-picker ${ className }` }>
      <Hue
        { ...styles.hue }
        hsl={ hsl }
        pointer={ pointer }
        onChange={ handleChange }
        direction={ direction }
      />
    </div>
  )
}

HuePicker.defaultProps = {
  width: '316px',
  height: '16px',
  direction: 'horizontal',
  pointer: HuePointer,
}

export default ColorWrap(HuePicker)
