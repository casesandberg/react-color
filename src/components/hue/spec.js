/* global test, expect */

import React from 'react'
import renderer from 'react-test-renderer'
import { red } from '../../helpers/color'

import Hue from './Hue'
import HuePointer from './HuePointer'

test('Hue renders correctly', () => {
  const tree = renderer.create(
    <Hue { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Hue renders vertically', () => {
  const tree = renderer.create(
    <Hue { ...red } width={ 20 } height={ 200 } direction="vertical" />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('HuePointer renders correctly', () => {
  const tree = renderer.create(
    <HuePointer />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
