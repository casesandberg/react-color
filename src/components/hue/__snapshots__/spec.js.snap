// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hue renders correctly 1`] = `
<div
  className="hue-picker "
  style={
    Object {
      "height": "16px",
      "position": "relative",
      "width": "316px",
    }
  }
>
  <div
    style={
      Object {
        "MozBorderRadius": "2px",
        "MozBoxShadow": undefined,
        "OBorderRadius": "2px",
        "OBoxShadow": undefined,
        "WebkitBorderRadius": "2px",
        "WebkitBoxShadow": undefined,
        "borderRadius": "2px",
        "bottom": "0px",
        "boxShadow": undefined,
        "left": "0px",
        "msBorderRadius": "2px",
        "msBoxShadow": undefined,
        "position": "absolute",
        "right": "0px",
        "top": "0px",
      }
    }
  >
    <div
      className="hue-horizontal"
      onMouseDown={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "height": "100%",
          "padding": "0 2px",
          "position": "relative",
        }
      }
    >
      <style>
        
                    .hue-horizontal {
                      background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                        33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                      background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                        17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                    }
        
                    .hue-vertical {
                      background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                        #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                      background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                        #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                    }
                  
      </style>
      <div
        style={
          Object {
            "left": "69.44444444444443%",
            "position": "absolute",
          }
        }
      >
        <div
          style={
            Object {
              "MozBorderRadius": "50%",
              "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "MozTransform": "translate(-9px, -1px)",
              "OBorderRadius": "50%",
              "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "OTransform": "translate(-9px, -1px)",
              "WebkitBorderRadius": "50%",
              "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "WebkitTransform": "translate(-9px, -1px)",
              "backgroundColor": "rgb(248, 248, 248)",
              "borderRadius": "50%",
              "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "height": "18px",
              "msBorderRadius": "50%",
              "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "msTransform": "translate(-9px, -1px)",
              "transform": "translate(-9px, -1px)",
              "width": "18px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Hue renders vertically 1`] = `
<div
  className="hue-picker "
  style={
    Object {
      "height": 200,
      "position": "relative",
      "width": 20,
    }
  }
>
  <div
    style={
      Object {
        "MozBorderRadius": "2px",
        "MozBoxShadow": undefined,
        "OBorderRadius": "2px",
        "OBoxShadow": undefined,
        "WebkitBorderRadius": "2px",
        "WebkitBoxShadow": undefined,
        "borderRadius": "2px",
        "bottom": "0px",
        "boxShadow": undefined,
        "left": "0px",
        "msBorderRadius": "2px",
        "msBoxShadow": undefined,
        "position": "absolute",
        "right": "0px",
        "top": "0px",
      }
    }
  >
    <div
      className="hue-vertical"
      onMouseDown={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "height": "100%",
          "padding": "0 2px",
          "position": "relative",
        }
      }
    >
      <style>
        
                    .hue-horizontal {
                      background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                        33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                      background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                        17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                    }
        
                    .hue-vertical {
                      background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                        #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                      background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                        #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                    }
                  
      </style>
      <div
        style={
          Object {
            "left": "0px",
            "position": "absolute",
            "top": "30.55555555555557%",
          }
        }
      >
        <div
          style={
            Object {
              "MozBorderRadius": "50%",
              "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "MozTransform": "translate(-3px, -9px)",
              "OBorderRadius": "50%",
              "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "OTransform": "translate(-3px, -9px)",
              "WebkitBorderRadius": "50%",
              "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "WebkitTransform": "translate(-3px, -9px)",
              "backgroundColor": "rgb(248, 248, 248)",
              "borderRadius": "50%",
              "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "height": "18px",
              "msBorderRadius": "50%",
              "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "msTransform": "translate(-3px, -9px)",
              "transform": "translate(-3px, -9px)",
              "width": "18px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`HuePointer renders correctly 1`] = `
<div
  style={
    Object {
      "MozBorderRadius": "50%",
      "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "MozTransform": "translate(-9px, -1px)",
      "OBorderRadius": "50%",
      "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "OTransform": "translate(-9px, -1px)",
      "WebkitBorderRadius": "50%",
      "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "WebkitTransform": "translate(-9px, -1px)",
      "backgroundColor": "rgb(248, 248, 248)",
      "borderRadius": "50%",
      "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "height": "18px",
      "msBorderRadius": "50%",
      "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "msTransform": "translate(-9px, -1px)",
      "transform": "translate(-9px, -1px)",
      "width": "18px",
    }
  }
/>
`;
